alf/utils/visualizer.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2021 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Various functions related to visualizations of networks etc."""
+
+import torch
+
+
+def critic_network_visualizer(net, observation, origin, dx, dy, H=20, W=20):
+    """Generate a batched network response image within the range specified by
+    ``origin``, ``dx``, ``dy`` as shown below:
+
+    ``origin`` ---> ``dx``
+        |
+        v
+      ``dy``
+
+    Example usage:
+
+    .. code-block:: python
+
+        # assume a case where the dimensionality of action is 4
+        # the upper-left point
+        origin = torch.Tensor([1, -1, 0, 0])
+        # the upper right point
+        dx = torch.Tensor([1, 1, 0, 0])
+        # the lower-left point
+        dy = torch.Tensor([-1, -1, 0, 0])
+
+        # define a network function
+        def net_func(net_input):
+            critics, _ = self._critic_networks(
+                net_input)  # [B, replicas * reward_dim]
+            critics = critics.reshape(  # [B, replicas, reward_dim]
+                -1, self._num_critic_replicas, *self._reward_spec.shape)
+            critics = critics.min(dim=1)[0]
+            return critics
+
+        img = critic_network_visualizer(net_func, inputs.observation,
+                                origin, dx, dy,
+                                20, 20)
+
+        # visualize the first response image in the batch
+        data = img[0, ...].squeeze(0)
+        data = data.cpu().numpy()
+
+        val_img = alf.summary.render.render_heatmap(name="val_img", data=data)
+
+
+    Args:
+        net (Callable): a callable that is called as``net((obsevation, actions))``
+        observation (Tensor): [B, ...]
+        origin (tensor): tensor representing the start of the probing region,
+            with the shape of [d]
+        dx (tensor): a tensor representing the end position of the probing
```

Comment:
If it's the end position, then probably shouldn't be called dx.

Reply:
Now changed the names to be more informative.